File: src/LeetCode/Algorithms/Symmetric_Tree.java
package LeetCode.Algorithms;

import org.w3c.dom.ls.LSOutput;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

// https://leetcode.com/problems/symmetric-tree/
public class Symmetric_Tree {
	public static class TreeNode {
		int val;
		TreeNode left;
		TreeNode right;
		
		TreeNode(int x) {
			this.val = x;
		}
	}
	
	public static void main(String[] args){
		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(2);
		root.right = new TreeNode(2);
		root.left.left = new TreeNode(3);
		root.left.right = new TreeNode(4);
		root.right.left = new TreeNode(4);
		root.right.right = new TreeNode(3);
		
		System.out.println(isSymmetric(root));
	}
	
	public static boolean isSymmetric(TreeNode root) {
		Queue<TreeNode> q = new LinkedList<>();
		
		if (root != null) {
			q.offer(root.left);
			q.offer(root.right);
			
			while (!q.isEmpty()) {
				TreeNode tn1 = q.poll();
				TreeNode tn2 = q.poll();
				
				if (tn1 == null && tn2 == null) {
					continue;
				}
				
				if (tn1 == null || tn2 == null) {
					return false;
				}
				
				if (tn1.val != tn2.val) {
					return false;
				}
				
				q.offer(tn1.left);
				q.offer(tn2.right);
				q.offer(tn1.right);
				q.offer(tn2.left);
			}
		}
		
		return true;
	}
}

File: src/LeetCode/Algorithms/JavaScript/Symmetric_Tree.js
/**
 * https://leetcode.com/problems/symmetric-tree/
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
	let queue = [];
	
	if (root) {
		queue.unshift(root.left);
		queue.unshift(root.right);
		
		while (queue.length > 0) {
			let curr1 = queue.pop();
			let curr2 = queue.pop();
			
			if (!curr1 && !curr2) {
				continue;
			}
			
			if ((!curr1 || !curr2) || (curr1.val !== curr2.val)) {
				return false;
			}
			
			queue.unshift(curr1.left);
			queue.unshift(curr2.right);
			queue.unshift(curr1.right);
			queue.unshift(curr2.left);
		}
	}
	
	return true;
};

function TreeNode(val) {
	this.val = val;
	this.left = this.right = null;
}

let root = new TreeNode(1);

/*root.left = new TreeNode(2);
root.right = new TreeNode(2);
root.left.left = new TreeNode(3);
root.left.right = new TreeNode(4);
root.right.left = new TreeNode(4);
root.right.right = new TreeNode(3);*/

root.left = new TreeNode(2);
root.right = new TreeNode(2);
root.left.right = new TreeNode(3);
root.right.right = new TreeNode(3);

console.log(isSymmetric(root));